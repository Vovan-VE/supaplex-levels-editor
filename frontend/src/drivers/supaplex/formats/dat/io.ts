import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { LEVEL_BYTES_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";

function* levelsFromBuffer(buffer: ArrayBuffer): Iterable<ISupaplexLevel> {
  if (buffer.byteLength % LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: not a module of level size");
  }

  const count = Math.floor(buffer.byteLength / LEVEL_BYTES_LENGTH);
  for (let i = 0; i < count; i++) {
    yield createLevel(
      LEVEL_WIDTH,
      LEVEL_HEIGHT,
      new Uint8Array(buffer, i * LEVEL_BYTES_LENGTH, LEVEL_BYTES_LENGTH),
    );
  }
}

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset =>
  createLevelset(levelsFromBuffer(file));

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  const result = new Uint8Array(levelset.levelsCount * LEVEL_BYTES_LENGTH);
  for (let [i, level] of levelset.getLevels().entries()) {
    // errors covered by "support report"
    if (level.width > LEVEL_WIDTH || level.height > LEVEL_HEIGHT) {
      throw new Error(
        `Level ${i + 1} is too large: ${level.width}x${level.height}`,
      );
    }

    // warnings covered by "support report"
    if (level.width < LEVEL_WIDTH || level.height < LEVEL_HEIGHT) {
      level = level.resize({ width: LEVEL_WIDTH, height: LEVEL_HEIGHT });
    }
    level = level.setDemo(null);

    result.set(level.raw, i * LEVEL_BYTES_LENGTH);
  }
  return result.buffer;
};
