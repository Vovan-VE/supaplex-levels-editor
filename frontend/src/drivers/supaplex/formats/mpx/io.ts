import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { FOOTER_BYTE_LENGTH } from "../std";

// MPX format
// 04 : "MPX "
// 04 : 01 00 01 00
//      -ver- -----
//            levels
//            count
// repeat for every level {
// 04 :  w 00  h 00
// 08 : 15 00 00 00
//      -offset+1--
// 0C : 82 06 00 00
//      ---size----
// }
// repeat for every level {
//    : std sp level with respect to w & h
//    : demo
// }
//
// - https://git.artsoft.org/?p=rocksndiamonds.git;a=commit;h=29ea60b3
// - https://github.com/ArtsoftEntertainment/rocksndiamonds/commit/29ea60b3#diff-6854d2255acd6155f61b1558b39cbce581fad371b897b8ab58a90e6a227a6a5eR370

const MPX_SIGN = new Uint8Array("MPX ".split("").map((ch) => ch.charCodeAt(0)));
const MIN_FILE_SIZE = 8 + 12 + 1 + FOOTER_BYTE_LENGTH;

const cmpUint8Array = (
  buffer: ArrayBuffer,
  byteOffset: number,
  match: Uint8Array,
) =>
  new Uint8Array(buffer, byteOffset, match.length).every(
    (b, i) => b === match[i],
  );

const getInt16LE = (file: ArrayBuffer, byteOffset: number) =>
  new DataView(file, byteOffset, 2).getInt16(0, true);

const getInt32LE = (file: ArrayBuffer, byteOffset: number) =>
  new DataView(file, byteOffset, 4).getInt32(0, true);

const setInt16LE = (buffer: Uint8Array, byteOffset: number, value: number) => {
  new DataView(buffer.buffer).setInt16(byteOffset, value, true);
};

const setInt32LE = (buffer: Uint8Array, byteOffset: number, value: number) => {
  new DataView(buffer.buffer).setInt32(byteOffset, value, true);
};

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset => {
  if (file.byteLength < MIN_FILE_SIZE || !cmpUint8Array(file, 0, MPX_SIGN)) {
    throw new Error("Unrecognized file format");
  }

  const ver = getInt16LE(file, 4);
  let levelsCount = getInt16LE(file, 6);
  const widths: number[] = [];
  const heights: number[] = [];
  const offsets: number[] = [];
  const sizes: number[] = [];
  if (ver === 0x2020 && levelsCount === 0x2020) {
    // 4 spaces
    //ver = 1;
    levelsCount = 1;
    widths.push(getInt16LE(file, 8));
    heights.push(getInt16LE(file, 10));
    offsets.push(20);
    sizes.push(widths[0] * heights[0] + FOOTER_BYTE_LENGTH);
  } else {
    if (ver !== 1) {
      throw new Error(`Unsupported format version ${ver}`);
    }
    for (let i = 0, at = 8; i < levelsCount; i++, at += 12) {
      widths.push(getInt16LE(file, at));
      heights.push(getInt16LE(file, at + 2));
      offsets.push(getInt32LE(file, at + 4) - 1);
      sizes.push(getInt32LE(file, at + 8));
    }
  }

  const levels: ISupaplexLevel[] = [];
  for (let i = 0; i < levelsCount; i++) {
    levels.push(
      createLevel(
        widths[i],
        heights[i],
        new Uint8Array(file, offsets[i], sizes[i]),
      ),
    );
  }

  return createLevelset(levels);
};

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  // warnings covered by "support report"
  const levelsCount = Math.min(0x7fff, levelset.levelsCount);
  const levels = levelset.getLevels().slice(0, levelsCount);

  const widths: number[] = [];
  const heights: number[] = [];
  const offsets: number[] = [];
  const sizes: number[] = [];

  let offset = 8 + 12 * levelsCount;
  for (const [i, level] of levels.entries()) {
    // errors covered by "support report"
    if (level.width > 0x7fff || level.height > 0x7fff) {
      throw new Error(
        `Level ${i + 1} is too large: ${level.width}x${level.height}`,
      );
    }
    if (offset >= 0x7fffffff - 1) {
      throw new Error(
        `Level ${
          i + 1
        } cannot be saved due to MPX file format limitations (offset+1 exceeds maximum)`,
      );
    }

    widths.push(level.width);
    heights.push(level.height);
    offsets.push(offset);
    sizes.push(level.length);
    offset += level.length;
  }

  const result = new Uint8Array(offset);
  result.set(MPX_SIGN, 0);
  setInt16LE(result, 4, 1);
  setInt16LE(result, 6, levelsCount);
  for (let i = 0, at = 8; i < levelsCount; i++, at += 12) {
    setInt16LE(result, at, widths[i]);
    setInt16LE(result, at + 2, heights[i]);
    setInt32LE(result, at + 4, offsets[i] + 1);
    setInt32LE(result, at + 8, sizes[i]);

    result.set(levels[i].raw, offsets[i]);
  }

  return result.buffer;
};
