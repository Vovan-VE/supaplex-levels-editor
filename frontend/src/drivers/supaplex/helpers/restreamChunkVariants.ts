import { ITilesStreamItem } from "../../types";

export function* restreamChunkVariants(
  chunk: ITilesStreamItem,
  getVariant: (x: number, y: number) => number | undefined,
): Iterable<ITilesStreamItem> {
  const [x, y, width, tile] = chunk;
  let lastItem: [
    x: number,
    y: number,
    width: number,
    tile: number,
    variant: number | undefined,
  ] = [x, y, 0, tile, undefined];
  for (let i = 0; i < width; i++) {
    const cx = x + i;
    const variant = getVariant(cx, y);
    if (variant === lastItem[4]) {
      lastItem[2]++;
    } else {
      if (lastItem[2]) {
        yield lastItem;
      }
      lastItem = [cx, y, 1, tile, variant];
    }
  }
  if (lastItem[2]) {
    yield lastItem;
  }
}
