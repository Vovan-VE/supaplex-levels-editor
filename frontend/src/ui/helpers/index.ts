export * from "./ClickOutside";
export * from "./Popper";
