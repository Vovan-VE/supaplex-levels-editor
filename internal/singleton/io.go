package singleton

import (
	"context"
	"fmt"
	"io"
	"net"
	"time"

	"github.com/pkg/errors"
	"github.com/wailsapp/wails/v2/pkg/logger"
)

var (
	ErrConnectNoDestination = errors.New("no destination")
)

const (
	name = "sple-ipc-v1"
)

func trySend(ctx context.Context, l logger.Logger, messages [][]byte) (done bool, err error) {
	ctxTimeout, cancel := context.WithTimeout(ctx, 1*time.Second)
	defer cancel()

	l.Debug("IPC connecting")
	conn, err := connect(ctxTimeout, name)
	if errors.Is(err, ErrConnectNoDestination) {
		l.Debug("IPC no primary accessible")
		return false, nil
	}
	if err != nil {
		l.Debug("IPC failed to connect")
		return false, errors.Wrap(err, "connect")
	}
	defer conn.Close()
	l.Debug("IPC connected")

	done = true
	for _, data := range messages {
		if len(data) > 0 {
			l.Debug(fmt.Sprintf("IPC send message: %s", string(data)))
			_, err = NewMessage(data).WriteTo(conn)
			if err != nil {
				return done, errors.Wrap(err, "Write")
			}
		}
	}
	return
}

func runReceiver(ctx context.Context, l logger.Logger, recv chan<- []byte) error {
	l.Debug("IPC start listen")
	listener, err := listen(ctx, name)
	if err != nil {
		l.Debug("IPC failed to listen")
		return errors.Wrap(err, "listen")
	}
	defer listener.Close()
	l.Debug("IPC listening ok")
	defer l.Debug("IPC listening end")

	done := make(chan error)
	go func() {
		defer close(done)
		done <- errors.Wrap(acceptLoop(ctx, listener, l, recv), "acceptLoop")
	}()
	return <-done
}

func acceptLoop(ctx context.Context, listener net.Listener, l logger.Logger, recv chan<- []byte) error {
	l.Debug("IPC accept loop")
	for {
		select {
		case <-ctx.Done():
			l.Debug("IPC accept context done")
			return ctx.Err()
		default:
		}

		l.Debug("IPC accept waiting a client")
		conn, err := listener.Accept()
		if err != nil {
			l.Debug("IPC accept failed")
			return errors.Wrap(err, "Accept")
		}
		l.Debug("IPC accepted a client")
		if err = conn.SetDeadline(time.Now().Add(1 * time.Second)); err != nil {
			return errors.Wrap(err, "SetDeadline")
		}
		if err = drainClient(ctx, conn, l, recv); err != nil {
			return errors.Wrap(err, "drainClient")
		}
	}
}

func drainClient(ctx context.Context, conn net.Conn, l logger.Logger, recv chan<- []byte) error {
	l.Debug("IPC draining a client")
	defer l.Debug("IPC end draining client")
	defer conn.Close()
	for {
		select {
		case <-ctx.Done():
			l.Debug("IPC draining context done")
			return ctx.Err()
		default:
		}

		l.Debug("IPC reading from client")
		msg, err := ReadMessage(conn)
		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
			l.Debug("IPC reading completed")
			return nil
		}
		if err != nil {
			l.Debug("IPC reading failed")
			return errors.Wrap(err, "read message")
		}
		l.Debug(fmt.Sprintf("IPC read message: %s", string(msg.Data)))
		recv <- msg.Data
	}
}
