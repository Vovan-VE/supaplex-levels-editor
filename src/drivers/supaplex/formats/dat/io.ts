import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { LEVEL_BYTES_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";

function* levelsFromBuffer(buffer: ArrayBuffer): Iterable<ISupaplexLevel> {
  if (buffer.byteLength % LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: not a module of level size");
  }

  const count = Math.floor(buffer.byteLength / LEVEL_BYTES_LENGTH);
  for (let i = 0; i < count; i++) {
    yield createLevel(
      LEVEL_WIDTH,
      LEVEL_HEIGHT,
      new Uint8Array(buffer, i * LEVEL_BYTES_LENGTH, LEVEL_BYTES_LENGTH),
    );
  }
}

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset =>
  createLevelset(levelsFromBuffer(file));

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  const result = new Uint8Array(levelset.levelsCount * LEVEL_BYTES_LENGTH);
  let i = 0;
  for (const level of levelset.getLevels()) {
    result.set(level.raw, i * LEVEL_BYTES_LENGTH);
    i++;
  }
  return result.buffer;
};
