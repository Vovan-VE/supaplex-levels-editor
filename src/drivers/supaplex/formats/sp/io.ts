import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevelset } from "../../types";
import { LEVEL_BYTES_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset => {
  if (file.byteLength < LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: less then level size");
  }

  let startOfSignature = LEVEL_BYTES_LENGTH;
  if (file.byteLength > LEVEL_BYTES_LENGTH) {
    const trailer = new Uint8Array(file, LEVEL_BYTES_LENGTH);
    const p = trailer.indexOf(0xff, 1);
    if (p > 0) {
      startOfSignature += p + 1;
    }
  }

  return createLevelset([
    createLevel(
      LEVEL_WIDTH,
      LEVEL_HEIGHT,
      new Uint8Array(file, 0, startOfSignature),
    ),
  ]);
};

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer =>
  levelset.getLevel(0).raw.buffer;
