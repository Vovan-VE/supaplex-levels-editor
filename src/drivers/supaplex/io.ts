import { ISupaplexLevelset, ISupaplexReader, ISupaplexWriter } from "./types";
import { LEVEL_BYTES_LENGTH, SupaplexLevel } from "./level";
import { SupaplexLevelset } from "./levelset";

function* levelsFromBuffer(buffer: ArrayBuffer) {
  if (buffer.byteLength % LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: not a module of level size");
  }

  const count = Math.floor(buffer.byteLength / LEVEL_BYTES_LENGTH);
  for (let i = 0; i < count; i++) {
    yield SupaplexLevel.fromArrayBuffer(buffer, i * LEVEL_BYTES_LENGTH);
  }
}

function* levelsToRaw(levelset: ISupaplexLevelset) {
  for (const level of levelset.getLevels()) {
    yield level.raw;
  }
}

export const readLevelset = async (file: Blob): Promise<ISupaplexLevelset> =>
  new SupaplexLevelset(levelsFromBuffer(await file.arrayBuffer()));

export const writeLevelset = async (
  levelset: ISupaplexLevelset,
): Promise<Blob> => new Blob([...levelsToRaw(levelset)]);

export const reader: ISupaplexReader = {
  readLevelset,
};

export const writer: ISupaplexWriter = {
  writeLevelset,
};
