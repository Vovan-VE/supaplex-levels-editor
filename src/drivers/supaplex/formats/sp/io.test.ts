import { dumpLevelset, readExampleFile } from "../../helpers.dev";
import { readLevelset, writeLevelset } from "./io";

it("read original levelset", async () => {
  const data = await readExampleFile("D95S018.sp");
  expect(data.length).toBe(0x6b0);

  const levelset = readLevelset(data.buffer);

  expect(dumpLevelset(levelset)).toMatchSnapshot("levelset");

  const result = writeLevelset(levelset);
  expect(result).toEqual(data.buffer.slice(0, 0x616));
});

it("invalid file size", () => {
  expect(() => readLevelset(new Uint8Array(100).buffer)).toThrow(
    new Error("Invalid file size: less then level size"),
  );
});
