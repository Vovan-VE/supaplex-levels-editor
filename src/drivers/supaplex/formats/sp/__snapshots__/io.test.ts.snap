// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read original levelset: levelset 1`] = `
Object {
  "levels": Array [
    Object {
      "cells": Array [
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▧573D·< ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓8 <· ⊏⊐^▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓6v⊏⊐ ·> ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓7 >· E ▧▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▧▓ ▓ 3▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓⊓ >@ ·> ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓⊔^▓DvD▓v·▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓0 <D D< ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓596·&·5▧▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓·▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
        "▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓",
      ],
      "demo": Uint8Array [
        100,
        0,
        227,
        25,
        38,
        8,
        180,
        240,
        226,
        240,
        48,
        241,
        17,
        132,
        0,
        241,
        1,
        0,
        180,
        240,
      ],
      "height": 24,
      "infotronsNeed": 1,
      "initialFreezeZonks": false,
      "initialGravity": false,
      "maxTitleLength": 23,
      "specialPorts": Array [],
      "title": "======= QUICKY! =======",
      "width": 60,
    },
  ],
  "levelsCount": 1,
}
`;
